hashicorp/terraform-provider-tfe
# Add ability to create Policy Set Versions in PolicySet Resource

## Description

This PR adds the ability to create Policy Set Versions and upload policies using the provider.

This change adds a new resource and a new data source.
1. `tfe_policy_set_version` is the new resource to create and upload policies to a policy set.
2. `tfe_slug` is used to manage local files. It tracks file changes and computes a checksum. It is currently minimal and primarily used by the `tfe_policy_set_version` to manage the local files.


#### Example usage

```
data "tfe_slug" "policy" {
  source_path = "policies/my-policy-set"
}

resource "tfe_policy_set_version" "test" {
  policy_set_id = tfe_policy_set.test.id
  policies_path_contents_checksum = data.tfe_slug.policy.checksum
  policies_path = data.tfe_slug.policy.source_path
}
```

## Testing plan

1. Pull down this branch and `make build` the provider. Make sure to enable [development override for the provider](https://www.terraform.io/docs/cli/config/config-file.html#development-overrides-for-provider-developers).
2. Create a new directory and `terraform init`
3. Copy the `tfe/test-fixtures/policy-set-version` directory to your local working directory under `policies/my-policy-set`
4. Add this to your config

```
resource "tfe_organization" "test" {
  name  = "test-terraform-rg"
  email = "admin@company.com"
}

resource "tfe_policy_set" "test" {
  name          = "my-policy-set"
  description   = "A brand new policy set"
  organization  = tfe_organization.test.id
}

data "tfe_slug" "policy" {
  source_path = "policies/my-policy-set"
}

resource "tfe_policy_set_version" "test" {
  policy_set_id = tfe_policy_set.test.id
  policies_path_contents_checksum = data.tfe_slug.policy.checksum
  policies_path = data.tfe_slug.policy.source_path
}

output "policy" {
  value = tfe_policy_set.test
}

output "policy_version" {
  value = tfe_policy_set_version.test
}

output "data_policy" {
  value = data.tfe_slug.policy
}
```

5. Run `terraform apply`
6. Validate in the UI in this organization that the policy set is created, and verify the terraform output that the `status` is `ready`. You can also [curl to get the PolicySetVersion and confirm the status](https://www.terraform.io/docs/cloud/api/policy-sets.html#show-a-policy-set-version).
7. Then edit/add a new file in the `policies/my-policy-set` directory and run `terraform apply` and notice a destroy/create take place. 

## Documentation

Images of the new documentation is put as a PR comment near the file.

## External links

- [API documentation](https://www.terraform.io/docs/cloud/api/policy-sets.html#upload-policy-set-versions)
- [go-tfe PR](https://github.com/hashicorp/go-tfe/pull/230)

Closes #279 

## Output from acceptance tests

```
$TFE_TOKEN=<redacted> TFE_HOSTNAME=tfe-zone-4c292d5b.ngrok.io TF_ACC=1 TF_LOG_PROVIDER=DEBUG  gotestsum --format testname -- ./tfe/... -run TestAccTFEPolicy
PASS tfe.TestAccTFEPolicySetVersionFiles_basic (4.18s)
PASS tfe.TestAccTFEPolicySetParameter_basic (7.95s)
PASS tfe.TestAccTFEPolicySetParameter_update (13.59s)
PASS tfe.TestAccTFEPolicySetParameter_import (8.76s)
PASS tfe.TestAccTFEPolicySet_basic (8.55s)
PASS tfe.TestAccTFEPolicySet_update (14.69s)
PASS tfe.TestAccTFEPolicySet_updateEmpty (13.40s)
PASS tfe.TestAccTFEPolicySet_updatePopulated (15.99s)
PASS tfe.TestAccTFEPolicySet_updateToGlobal (14.26s)
PASS tfe.TestAccTFEPolicySet_updateToWorkspace (14.74s)
PASS tfe.TestAccTFEPolicySet_invalidName (0.77s)
PASS tfe.TestAccTFEPolicySetImport (8.98s)
PASS tfe.TestAccTFEPolicySetVersion_basic (7.27s)
PASS tfe.TestAccTFEPolicySetVersion_recreate (12.72s)
PASS tfe

=== Skipped
=== SKIP: tfe TestAccTFEPolicySet_vcs (0.37s)
    resource_tfe_policy_set_test.go:283: Please set GITHUB_TOKEN to run this test

=== SKIP: tfe TestAccTFEPolicySet_updateVCSBranch (0.36s)
    resource_tfe_policy_set_test.go:333: Please set GITHUB_TOKEN to run this test

DONE 16 tests, 2 skipped in 149.439s
```

## Other Considerations

One consideration was to use the [archive provider](https://registry.terraform.io/providers/hashicorp/archive/latest/docs/data-sources/archive_file) instead of the new data source, but having the user have to reference a different provider is a worse user experience than reference a tfe-provider native data source.
